[SRVKS-1306] Configuring long running requests (#89935)
openshift/openshift-docs: modules/configuring-default-route-timeouts.adoc

Reviewer: For these longer commands, I would suggest breaking them across lines so more of the command is easy to see without scrolling

You can configure the timeouts of a specific revision by updating your `KnativeServing` custom resource (CR), the Service definition, and using the `serving.knative.openshift.io/setRouteTimeout` annotation to adjust the {ocp-product-title} route timeout.

.Procedure

. Set the `max-revision-timeout` annotation in your `KnativeServing` CR as you require:
+
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    defaults:
      max-revision-timeout-seconds: "900"
----

. Optional: Verify the termination grace period of an activator by running the following command:
+
[source,terminal]
----
$ oc get deployment activator -n knative-serving -o jsonpath="{.spec.template.spec.terminationGracePeriodSeconds}"

900
----

. If necessary for your cloud provider, adjust the load balancer timeout by running the following command:
+
.Load balancer timeout adjustment for AWS Classic LB 
[source,terminal]
----

> $ oc -n openshift-ingress-operator patch ingresscontroller/default

Reviewer 2: See other comment about how to break up long commands.

// Module included in the following assemblies:
//
// * knative-serving/config-applications/configuring-revision-timeouts.adoc
:_content-type: PROCEDURE
[id="configuring-default-route-timeouts_{context}"]
= Configuring the default route timeouts per revision

> By configuring the route timeouts per revision, you can fine-tune timeouts for workloads with unique requirements, such as AI or data processing applications, without impacting the global timeout settings for other services.

Reviewer 3: ```suggestion
By configuring the route timeouts per revision, you can fine-tune timeouts for workloads with unique requirements, such as AI or data processing applications, without impacting the global timeout settings for other services.
You can configure the timeouts of a specific revision by updating your `KnativeServing` custom resource (CR), the Service definition, and using the `serving.knative.openshift.io/setRouteTimeout` annotation to adjust the OCP route timeout.
```

.Procedure

> You can configure the timeouts of a specific revision by updating your `KnativeServing` custom resource (CR), the Service definition, and using the `serving.knative.openshift.io/setRouteTimeout` annotation to adjust the OCP route timeout.

Reviewer 3: ```suggestion

```

> . Set the `max-revision-timeout` as required:

Reviewer 3: ```suggestion
. Set the `max-revision-timeout` annotation as required:
```

Reviewer 3: When you say `as required` do you mean that the user should know the value of this and set it according to their cluster's requirements. If so, I'd maybe reword to something like `Set the ... annotation to fit your requirements` or `as you require`.

+
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    defaults:
      max-revision-timeout-seconds: "900"
----
+
You can verify the activator's termination grace period: 
+
[source,terminal]
----
$ oc get deployment activator -n knative-serving -o jsonpath="{.spec.template.spec.terminationGracePeriodSeconds}"
900
----

. Set the timeout for your specific service:
+
[source,yaml]
----
apiVersion: serving.knative.dev/v1f
kind: Service
metadata:

>   name:<>

Reviewer 3: ```suggestion
  name: <my_service>
```

  annotations:
    serving.knative.openshift.io/setRouteTimeout: "800" <1>
spec:
  template:
    metadata:
      annotations:

> #... 

Reviewer 3: ```suggestion
#...
```

    spec:
      timeoutSeconds: 800 <2>
      responseStartTimeoutSeconds: 800 <3>
----
<1> This annotation sets the timeout for the OCP route. You can fine-tune this for each service instead of setting a global maximum.
<2> This ensures that the request will not exceed the specific value.
<3> This ensures that the response start timeout does not trigger before the max threshold is reached. The default value is `300`.
+

> . Depending on your cloud provider, you may also need to set or adjust the load balancer timeout:

Reviewer 3: Same as previous comment on line 70.

+
.Example of load balancer timeout adjustment for AWS Classic LB 
[source,terminal]
----
$ oc -n openshift-ingress-operator patch ingresscontroller/default --type=merge --patch='{"spec":{"endpointPublishingStrategy": {"type":"LoadBalancerService", "loadBalancer": {"scope":"External", "providerParameters":{"type":"AWS", "aws": {"type":"Classic", "classicLoadBalancer": {"connectionIdleTimeout":"20m"}}}}}}}'

> ingresscontroller.operator.openshift.io/default patched

Reviewer 3: ```suggestion

```

----

> 

Reviewer 3: ```suggestion
+
```

[IMPORTANT]
====

> When using {SMProductShortName}, if the activator pod is stopped while a long-running request is in-flight, the request will be interrupted.

Reviewer 3: Same comments as the previous module.

// Module included in the following assemblies:
//
// * knative-serving/config-applications/configuring-revision-timeouts.adoc
:_content-type: PROCEDURE
[id="configuring-default-route-timeouts_{context}"]
= Configuring the default route timeouts per revision

By configuring the route timeouts per revision, you can fine-tune timeouts for workloads with unique requirements, such as AI or data processing applications, without impacting the global timeout settings for other services.

> You can configure the timeouts of a specific revision by updating your `KnativeServing` custom resource (CR), the Service definition, and using the `serving.knative.openshift.io/setRouteTimeout` annotation to adjust the OCP route timeout.

Reviewer 4: ```suggestion
You can configure the timeouts of a specific revision by updating your `KnativeServing` custom resource (CR), the `Service` definition, and using the `serving.knative.openshift.io/setRouteTimeout` annotation to adjust the OpenShift route timeout.
```

Should it be "OpenShift route timeout."? We don't use OCP though, you can use the usual openshift attribute that is used in serveless

.Procedure

. Set the `max-revision-timeout` annotation in your `KnativeServing` CR as you require:
+
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    defaults:
      max-revision-timeout-seconds: "900"
----
+

> You can verify the activator's termination grace period by running the following command:

Reviewer 4: ```suggestion
You can verify the termination grace period of an activator by running the following command:
```

Same comment -- remove `'s`

+
[source,terminal]
----
$ oc get deployment activator -n knative-serving -o jsonpath="{.spec.template.spec.terminationGracePeriodSeconds}"
900
----

. Optional: Adjust the load balancer timeout if necessary for your cloud provider:
+
.Example of load balancer timeout adjustment for AWS Classic LB 
[source,terminal]
----
$ oc -n openshift-ingress-operator patch ingresscontroller/default --type=merge --patch='{"spec":{"endpointPublishingStrategy": {"type":"LoadBalancerService", "loadBalancer": {"scope":"External", "providerParameters":{"type":"AWS", "aws": {"type":"Classic", "classicLoadBalancer": {"connectionIdleTimeout":"20m"}}}}}}}'
----

. Set the timeout for your specific service:
+
[source,yaml]
----
apiVersion: serving.knative.dev/v1f
kind: Service
metadata:
  name: <your_service_name>
  annotations:
    serving.knative.openshift.io/setRouteTimeout: "800" <1>
spec:
  template:
    metadata:
      annotations:
#...
    spec:
      timeoutSeconds: 800 <2>
      responseStartTimeoutSeconds: 800 <3>
----

> <1> This annotation sets the timeout for the OCP route. You can fine-tune this for each service instead of setting a global maximum.

Reviewer 4: ```suggestion
<1> This annotation sets the timeout for the OpenShift route. You can fine-tune this for each service instead of setting a global maximum.
```

Same comment about removing "OCP"

> <2> This ensures that the request will not exceed the specific value.

Reviewer 4: ```suggestion
<2> This ensures that the request does not exceed the specific value.
```

<3> This ensures that the response start timeout does not trigger before the max threshold is reached. The default value is `300`.

+
[IMPORTANT]
====

> When using {SMProductShortName}, if the activator pod is stopped while a long-running request is in-flight, the request will be interrupted.

Reviewer 4: ```suggestion
When using {SMProductShortName}, if the activator pod is stopped while a long-running request is in-flight, the request is interrupted.
```

> To avoid request interruptions, you must adjust the `terminationDrainDuration` value in the `ServiceMeshControlPlane`: 

Reviewer 4: ```suggestion
To avoid request interruptions, you must adjust the `terminationDrainDuration` value in the `ServiceMeshControlPlane` CR: 
```

[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
#...
spec: 
  techPreview:
      meshConfig:
        defaultConfig:
          terminationDrainDuration: 1000s <1>
#... 
----
<1> Ensure that the value exceeds the request duration to avoid the Istio proxy shutdown, which would interrupt the request.
====

.Verification

> If you are using Kourier, you can verify the current value of the timeout at the {ocp-product-title} route by running the following command:

Reviewer 4: ```suggestion
If you are using Kourier, verify the current value of the timeout at the {ocp-product-title} route by running the following command:
```

Just noticed this {ocp-product-title}  attribute here. Please use the attribute throughout the PR instead of using "OCP". (see comments earlier).

Author: Thanks for pointing that out! I used 'OCP' while drafting and accidentally left it in.

// Module included in the following assemblies:
//

> // * knative-serving/config-applications/configuring-revision-timeouts.adoc
> :_mod-docs-content-type: PROCEDURE

Reviewer: ```suggestion
// * knative-serving/config-applications/configuring-revision-timeouts.adoc

:_mod-docs-content-type: PROCEDURE
```

[id="configuring-default-route-timeouts_{context}"]
= Configuring the default route timeouts per revision

By configuring the route timeouts per revision, you can fine-tune timeouts for workloads with unique requirements, such as AI or data processing applications, without impacting the global timeout settings for other services.
You can configure the timeouts of a specific revision by updating your `KnativeServing` custom resource (CR), the Service definition, and using the `serving.knative.openshift.io/setRouteTimeout` annotation to adjust the {ocp-product-title} route timeout.

.Procedure

. Set the `max-revision-timeout` annotation in your `KnativeServing` CR as you require:
+
[source,yaml]
----
apiVersion: operator.knative.dev/v1beta1
kind: KnativeServing
metadata:
  name: knative-serving
spec:
  config:
    defaults:
      max-revision-timeout-seconds: "900"
----

. Optional: Verify the termination grace period of an activator by running the following command:
+
[source,terminal]
----
$ oc get deployment activator -n knative-serving -o jsonpath="{.spec.template.spec.terminationGracePeriodSeconds}"
900
----

> . If necessary for your cloud provider, adjust the load balancer timeout:

Reviewer: ```suggestion
. If necessary for your cloud provider, adjust the load balancer timeout by running the following command:
```
